lib/shared/s3-packages.js

```diff
@@ -0,0 +1,182 @@
+
+'use strict';
+
+const _ = require('lodash');
+const semver = require('semver');
+const Bluebird = require('bluebird');
+const request = Bluebird.promisifyAll(require('request'));
+const xml = Bluebird.promisifyAll(require('xml2js'));
+
+/**
+ * @summary Etcher production releases S3 bucket URL
+ * @constant
+ * @public
+ * @type {String}
+ */
+exports.BUCKET_URL_PRODUCTION = 'https://resin-production-downloads.s3.amazonaws.com';
+
+/**
+ * @summary Etcher releases S3 package name
+ * @constant
+ * @private
+ * @type {String}
+ */
+const S3_PACKAGE_NAME = 'etcher';
+
+/**
+ * @summary Number of packages per Etcher version
+ * @constant
+ * @private
+ * @type {Number}
+ */
+const NUMBER_OF_PACKAGES = 8;
+
+/**
+ * @summary Find the latest version from an S3 bucket XML representation
+ * @function
+ * @private
+ *
+ * @param {String} bucketXML - s3 bucket XML
+ * @param {String} [range] - semver range
+ * @fulfil {String} - latest version
+ * @returns {Promise}
+ *
+ * request(s3Packages.BUCKET_URL_PRODUCTION, (error, response, body) => {
+ *   if (error) {
+ *     throw error;
+ *   }
+ *
+ *   s3Packages.getLatestVersionFromBucketXML(body, '>=2.5.0').then((latestVersion) => {
+ *     console.log(`The latest version is: ${latestVersion}`);
+ *   });
+ * });
+ */
+exports.getLatestVersionFromBucketXML = (bucketXML, range = '*') => {
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return xml.parseStringAsync(bucketXML)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('ListBucketResult')
+    .get('Contents')
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      // The `semver` module refuses to apply ranges to prerelease versions
+      // As a workaround, we drop the prerelease tags, if any, apply the range
+      // on that, and keep using the prerelease tag from then on.
+      // See https://github.com/npm/node-semver#prerelease-tags
+      const strippedVersion = `${semver.major(version)}.${semver.minor(version)}.${semver.patch(version)}`;
+
+      if (name === S3_PACKAGE_NAME && semver.satisfies(strippedVersion, range)) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+
+        if (accumulator[version] >= NUMBER_OF_PACKAGES
+            && semver.gt(version, accumulator.latest)) {
+          accumulator.latest = version;
+        }
+      }
+
+      return accumulator;
+    }, {
+
+      // This is a good accumulator default value since
+      // every version is semantically greater than this.
+      latest: '0.0.0'
+
+    })
+    .get('latest');
+};
+
+/**
+ * @summary Query the latest available version from an S3 bucket
+ * @function
+ * @private
+ *
+ * @param {String} bucketUrl - an s3 bucket url
+ * @param {Object} options - options
+ * @param {String} options.currentVersion - current version
+ * @param {String} [options.range] - semver range
```

Comment:
I think it should be something internal to this module, and its unlikely to change unless we add a new deployment target.